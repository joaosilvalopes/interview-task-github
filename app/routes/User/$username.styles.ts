import { Link } from '@remix-run/react';
import styled from 'styled-components';
import Spinner from '~/components/Spinner';

export const UserContainer = styled.main` 
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  min-height: calc(100vh - ${props => props.theme.headerHeight});
  color: ${props => props.theme.colors.main};
  background-color: ${props => props.theme.colors.mainBackground};
`;

export const UserHeader = styled.div`
  display: flex;
  align-items: center;
  flex-direction: column;
  width: 100%;
  padding: 1rem;
  background-color: ${props => props.theme.colors.userHeaderBackground};

  ${props => props.theme.media.gteSm} {
    flex-direction: row;
  }
`;

export const UserAvatar = styled.img`
  width: 5rem;
  height: 5rem;
  border-radius: 50%;
  box-shadow: 0px 2px 2px rgba(0, 0, 0, 0.2);

  ${props => props.theme.media.gteSm} {
    margin-right: 1rem;
  }
`;

export const UserName = styled.h1`
  font-size: 2rem;
  color: ${props => props.theme.colors.username};
  margin: 1rem 0;

  ${props => props.theme.media.gteSm} {
    font-size: 2.5rem;
  }

  ${props => props.theme.media.gteMd} {
    font-size: 3rem;
  }
`;

export const UserRepositoriesCount = styled.p`
  font-size: 1.2rem;
  color: ${props => props.theme.colors.userRepositoriesCount};

  ${props => props.theme.media.gteSm} {
    margin-left: auto;
  }
`;

export const UserRepositoriesList = styled.ul`
  display: flex;
  flex-direction: column;
  width: 100%;
  height: 30rem;
`;

export const UserRepository = styled.li`
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  height: 20%;
  width: 100%;
  padding: 0.75rem;
  border-bottom: 1px solid ${props => props.theme.colors.userRepositoryBorder};

  &:first-child {
    border-top: 1px solid ${props => props.theme.colors.userRepositoryBorder};
  }
`;

export const UserRepositoryName = styled.h2`
  font-size: 2rem;
  color: ${props => props.theme.colors.userRepositoryName};
`;

export const UserRepositoryDescription = styled.p`
  font-size: 1.2rem;
  color: ${props => props.theme.colors.userRepositoriesDescription};
  white-space: nowrap;
  overflow-x: clip;
  text-overflow: ellipsis;
`;

export const Pagination = styled.div`
  display: flex;
  width: 100%;
  justify-content: center;
  margin-top: 1rem;
`;

export const PaginationButton = styled.button`
  padding: 0.5rem 1rem;
  font-size: 1.2rem;
  border: none;
  border-radius: 0.5rem;
  color: ${props => props.theme.colors.userPaginationButton};
  background-color: ${props => props.theme.colors.userPaginationButtonBackground};
  cursor: pointer;
  transition: background-color 0.2s ease-in-out;

  &:not(:disabled):hover {
    background-color: ${props => props.theme.colors.userPaginationButtonBackgroundOnHover};
  }

  &:disabled {
    cursor: wait;
  }
`;

export const BackLink = styled(Link)`
  display: block;
  margin: auto 0 1rem 0.75rem;
  padding: 0.5rem 1rem;
  border-radius: 0.5rem;
  cursor: pointer;
  font-size: 1.2rem;
  text-decoration: none;
  transition: background-color 0.2s ease-in-out;
  color: ${props => props.theme.colors.backButton};
  background-color: ${props => props.theme.colors.backButtonBackground};

  &:hover {
    background-color: ${props => props.theme.colors.backButtonBackgroundOnHover};
  }
`;

export const PageCount = styled.span`
  display: flex;
  align-items: center;
  margin: 0 1rem;
  color: ${props => props.theme.colors.pageCount};
  font-size: 1.2rem;
`;

export const PageSpinner = styled(Spinner)`
  font-size: 10rem;
  border-width: 0.5rem;
  margin: auto;
  color: ${props => props.theme.colors.pageSpinner};
`;
