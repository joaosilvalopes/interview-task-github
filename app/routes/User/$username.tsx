import type { LoaderArgs } from '@remix-run/node';
import { useLoaderData } from '@remix-run/react';
import Spinner from '~/components/Spinner';
import usePagination from '~/hooks/usePagination';

import type { Repository } from '~/sdk/getRepositories';
import getRepositories from '~/sdk/getRepositories';
import type { User } from '~/sdk/getUser';
import getUser from '~/sdk/getUser';

import {
  BackLink,
  Pagination,
  PaginationButton,
  UserAvatar,
  UserHeader,
  UserName,
  UserRepositoriesCount,
  UserRepositoriesList,
  UserRepository,
  UserRepositoryDescription,
  UserRepositoryName,
  PageCount,
  UserContainer,
  PageSpinner
} from './$username.styles';

type UserLoaderArgs = LoaderArgs & {
  params: {
    username: string;
  }
};

type ServerData = {
  user: User;
  repositories: Repository[];
};

export const PAGE_SIZE = 5;

export const loader = async ({ params }: UserLoaderArgs) => {
  const { username } = params;

  try {
    const [user, repositories] = await Promise.all([
      getUser(username),
      getRepositories(username, 1, PAGE_SIZE),
    ]);

    return { user, repositories };
  } catch (e) {
    console.error(e);
  }
};

const UserPage = () => {
  const { user, repositories } = useLoaderData<ServerData>();
  const fetchPage = (page: number) => getRepositories(user.username, page, PAGE_SIZE);

  const {
    pageEntries: visibleRepos,
    isPageLoading,
    page,
    pageCount,
    nextPage,
    previousPage
  } = usePagination<Repository>({
    pageSize: PAGE_SIZE,
    firstPageEntries: repositories,
    entryCount: user.repositoriesCount,
    fetchPage
  });

  return (
    <UserContainer>
      <UserHeader>
        <UserAvatar src={user.avatarUrl} data-testid="user-avatar" />
        <UserName data-testid="user-name">{user.name}</UserName>
        <UserRepositoriesCount data-testid="user-repos-count">Found {user.repositoriesCount} repositories</UserRepositoriesCount>
      </UserHeader>
      <UserRepositoriesList>
        {isPageLoading ? <PageSpinner data-testid="loading-page-indicator" /> : visibleRepos.map(repo => (
          <UserRepository key={repo.id}>
            <UserRepositoryName data-testid={`repo-name-${repo.id}`}>{repo.name}</UserRepositoryName>
            {repo.description && <UserRepositoryDescription data-testid={`repo-description-${repo.id}`}>{repo.description}</UserRepositoryDescription>}
          </UserRepository>
        ))}
      </UserRepositoriesList>
      <Pagination>
        {page > 1 && (
          <PaginationButton data-testid="previous-page-button" onClick={previousPage}>
            Previous
          </PaginationButton>
        )}
        <PageCount data-testid="page-count">{page}/{pageCount}</PageCount>
        {page < pageCount && (
          <PaginationButton data-testid="next-page-button" disabled={isPageLoading} onClick={nextPage}>
            { isPageLoading ? <Spinner /> : 'Next' } 
          </PaginationButton>
        )}
      </Pagination>
      <BackLink data-testid="back-link" to="/">Back</BackLink>
    </UserContainer>
  );
};

export default UserPage;
