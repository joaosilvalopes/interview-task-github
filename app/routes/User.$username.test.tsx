import { fireEvent, render, waitFor } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';

import User from './User.$username';

const pageSize = 10;

const user = {
    username: 'testuser',
    name: 'Test User',
    avatarUrl: 'https://example.com/avatar.png',
    repositoriesCount: pageSize * 3
};

const repositories = Array.from({ length: user.repositoriesCount }, (_, i) => ({
    id: i + 1,
    name: `Repository ${i + 1}`,
    description: `Description of Repository ${i + 1}`
}));

jest.mock('~/sdk/getRepositories', () => jest.fn((_, page: number) => repositories.slice((page - 1) * pageSize, page * pageSize)));

jest.mock('@remix-run/react', () => ({
    ...jest.requireActual('@remix-run/react'),
    useLoaderData: jest.fn(() => ({
        user,
        repositories: repositories.slice(0, pageSize),
    })),
}));

const renderWithRouter = () => render(<Router><User /></Router>)

describe('UserPage component', () => {
    it('renders user information and repositories', async () => {
        const { getByTestId, queryByTestId, asFragment } = renderWithRouter();

        expect(getByTestId('user-avatar')).toHaveAttribute("src", user.avatarUrl);
        expect(getByTestId('user-name')).toHaveTextContent(user.name);
        expect(getByTestId('user-repos-count')).toHaveTextContent(`Total number of repositories: ${user.repositoriesCount}`);

        for(const repo of repositories.slice(0, pageSize)) {
            expect(getByTestId(`repo-name-${repo.id}`)).toHaveTextContent(repo.name);
            expect(getByTestId(`repo-description-${repo.id}`)).toHaveTextContent(repo.description);
        }

        expect(queryByTestId(`loading-page-indicator`)).not.toBeInTheDocument();
        expect(getByTestId('page-count')).toHaveTextContent(`1/${Math.ceil(user.repositoriesCount / pageSize)}`);
        expect(queryByTestId(`previous-page-button`)).not.toBeInTheDocument();
        expect(getByTestId(`next-page-button`)).toBeInTheDocument();

        const nextPageButton = getByTestId('next-page-button');
    
        fireEvent.click(nextPageButton);

        expect(getByTestId(`loading-page-indicator`)).toBeInTheDocument();

        await waitFor(() => {
            for(const repo of repositories.slice(pageSize, pageSize * 2)) {
                expect(getByTestId(`repo-name-${repo.id}`)).toHaveTextContent(repo.name);
                expect(getByTestId(`repo-description-${repo.id}`)).toHaveTextContent(repo.description);
            }
        })

        expect(queryByTestId(`loading-page-indicator`)).not.toBeInTheDocument();
        expect(getByTestId('page-count')).toHaveTextContent(`2/${Math.ceil(user.repositoriesCount / pageSize)}`);
        expect(getByTestId(`previous-page-button`)).toBeInTheDocument();
        expect(getByTestId(`next-page-button`)).toBeInTheDocument();

        fireEvent.click(nextPageButton);

        expect(getByTestId(`loading-page-indicator`)).toBeInTheDocument();

        await waitFor(() => {
            for(const repo of repositories.slice(pageSize * 2, pageSize * 3)) {
                expect(getByTestId(`repo-name-${repo.id}`)).toHaveTextContent(repo.name);
                expect(getByTestId(`repo-description-${repo.id}`)).toHaveTextContent(repo.description);
            }
        })

        expect(queryByTestId(`loading-page-indicator`)).not.toBeInTheDocument();
        expect(getByTestId('page-count')).toHaveTextContent(`3/${Math.ceil(user.repositoriesCount / pageSize)}`);
        expect(getByTestId(`previous-page-button`)).toBeInTheDocument();
        expect(queryByTestId(`next-page-button`)).not.toBeInTheDocument();

        expect(asFragment()).toMatchSnapshot();
    });
});
