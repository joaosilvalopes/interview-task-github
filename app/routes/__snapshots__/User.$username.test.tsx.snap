// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserPage component renders user information and repositories 1`] = `
<DocumentFragment>
  <header
    class="sc-beySbM kYDCsI"
  >
    <img
      class="sc-guDLRT dVWhez"
      data-testid="user-avatar"
      src="https://example.com/avatar.png"
    />
    <h1
      class="sc-dmyDGi fRGVKL"
      data-testid="user-name"
    >
      Test User
    </h1>
    <p
      class="sc-hLQRIN jvyuRi"
      data-testid="user-repos-count"
    >
      Total number of repositories: 30
    </p>
  </header>
  <ul
    class="sc-eDLJxc eurhjQ"
  >
    <li
      class="sc-jTQDnj RMjSV"
    >
      <h2
        class="sc-gLLvby djhHsJ"
        data-testid="repo-name-21"
      >
        Repository 21
      </h2>
      <p
        class="sc-iBdmCd hmNQKR"
        data-testid="repo-description-21"
      >
        Description of Repository 21
      </p>
    </li>
    <li
      class="sc-jTQDnj RMjSV"
    >
      <h2
        class="sc-gLLvby djhHsJ"
        data-testid="repo-name-22"
      >
        Repository 22
      </h2>
      <p
        class="sc-iBdmCd hmNQKR"
        data-testid="repo-description-22"
      >
        Description of Repository 22
      </p>
    </li>
    <li
      class="sc-jTQDnj RMjSV"
    >
      <h2
        class="sc-gLLvby djhHsJ"
        data-testid="repo-name-23"
      >
        Repository 23
      </h2>
      <p
        class="sc-iBdmCd hmNQKR"
        data-testid="repo-description-23"
      >
        Description of Repository 23
      </p>
    </li>
    <li
      class="sc-jTQDnj RMjSV"
    >
      <h2
        class="sc-gLLvby djhHsJ"
        data-testid="repo-name-24"
      >
        Repository 24
      </h2>
      <p
        class="sc-iBdmCd hmNQKR"
        data-testid="repo-description-24"
      >
        Description of Repository 24
      </p>
    </li>
    <li
      class="sc-jTQDnj RMjSV"
    >
      <h2
        class="sc-gLLvby djhHsJ"
        data-testid="repo-name-25"
      >
        Repository 25
      </h2>
      <p
        class="sc-iBdmCd hmNQKR"
        data-testid="repo-description-25"
      >
        Description of Repository 25
      </p>
    </li>
    <li
      class="sc-jTQDnj RMjSV"
    >
      <h2
        class="sc-gLLvby djhHsJ"
        data-testid="repo-name-26"
      >
        Repository 26
      </h2>
      <p
        class="sc-iBdmCd hmNQKR"
        data-testid="repo-description-26"
      >
        Description of Repository 26
      </p>
    </li>
    <li
      class="sc-jTQDnj RMjSV"
    >
      <h2
        class="sc-gLLvby djhHsJ"
        data-testid="repo-name-27"
      >
        Repository 27
      </h2>
      <p
        class="sc-iBdmCd hmNQKR"
        data-testid="repo-description-27"
      >
        Description of Repository 27
      </p>
    </li>
    <li
      class="sc-jTQDnj RMjSV"
    >
      <h2
        class="sc-gLLvby djhHsJ"
        data-testid="repo-name-28"
      >
        Repository 28
      </h2>
      <p
        class="sc-iBdmCd hmNQKR"
        data-testid="repo-description-28"
      >
        Description of Repository 28
      </p>
    </li>
    <li
      class="sc-jTQDnj RMjSV"
    >
      <h2
        class="sc-gLLvby djhHsJ"
        data-testid="repo-name-29"
      >
        Repository 29
      </h2>
      <p
        class="sc-iBdmCd hmNQKR"
        data-testid="repo-description-29"
      >
        Description of Repository 29
      </p>
    </li>
    <li
      class="sc-jTQDnj RMjSV"
    >
      <h2
        class="sc-gLLvby djhHsJ"
        data-testid="repo-name-30"
      >
        Repository 30
      </h2>
      <p
        class="sc-iBdmCd hmNQKR"
        data-testid="repo-description-30"
      >
        Description of Repository 30
      </p>
    </li>
  </ul>
  <div
    class="sc-fsYeqs kLqGoS"
  >
    <button
      class="sc-qZqnM jbOvBn"
      data-testid="previous-page-button"
    >
      Previous
    </button>
    <span
      class="sc-kFCsca hANyfT"
      data-testid="page-count"
    >
      3/3
    </span>
  </div>
  <a
    class="sc-jsEegt wKvXi"
    href="/"
  >
    Go back to Root route
  </a>
</DocumentFragment>
`;
